cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestBase.java

```diff
@@ -111,12 +115,18 @@ public void setUp() throws Exception {
       doClear(configuredNamespace, false);
     }
     registeredNamespaces.put(configuredNamespace, deleteUponTeardown);
+    applicationManagers.clear();
     LOG.info("Completed setUp.");
   }
 
   @After
   public void tearDown() throws Exception {
     LOG.info("Beginning tearDown.");
+
+    // Stop any running programs created in @Test
+    for (ApplicationManager manager: applicationManagers) {
```

Comment:
The service keeps retrying if it fails to fetch TMS notifications, so since it's always running, yes there are technically infinite retries on failure.